rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationServiceImpl.scala

```diff
@@ -369,17 +193,23 @@ class NodeConfigurationServiceImpl(
     }
   }
 
-  override def detectChangeInNodes(nodes : Seq[NodeConfiguration], cache: Map[NodeId, NodeConfigurationCache], directiveLib: FullActiveTechniqueCategory) : Set[RuleId] = {
-    nodes.flatMap{ x =>
-      detectChangeInNode(cache.get(x.nodeInfo.id), x, directiveLib, cache.size>0)
-    }.toSet
+  override def detectChangeInNodes(nodes : Seq[NodeConfiguration], cache: Map[NodeId, NodeConfigurationHash], directiveLib: FullActiveTechniqueCategory) : Set[RuleId] = {
+
+    // changes in nodes don't imply changes in directives / rules / serial.
+    // these change will lead to a new config id and generation if needed.
+    // we only increment serial when the cache is empty, which mean that it
+    // is a new rudder, or a "clear cache".
+    // Note that if the cache is empty, all nodes will be rewriten in all
+    // cases, but that allows to check that nodes really get their
+    // policies (because node config id may be the same, but they will continue sending
+    // reports with the previous serial)
```

Comment:
given the comment and the code, the name of the method is super-confusing
Could it be "initializeNodeCache" or something like that ?
Plus ,the directiveLib isn't used at all, you can remove it from the signature